File: funcscript/parser/syntax/FuncScriptParser.GetCallAndMemberAccess.cs
using FuncScript.Block;
namespace FuncScript.Core
{
    public partial class FuncScriptParser
    {
        static ExpressionBlockResult GetCallAndMemberAccess(ParseContext context, int index)
        {
            ExpressionBlock prog = null;
            var i1 = SkipSpace(context, index).NextIndex;
            var (theUnit, parseNode, i) = GetUnit(context, i1);
            if (i == index)
                return new ExpressionBlockResult(prog, parseNode, index);

            do
            {
                var (funcCall, nodeParList, i2) = GetFunctionCallParametersList(context, theUnit, i);
                if (i2 > i)
                {
                    i = i2;
                    theUnit = funcCall;
                    theUnit.CodePos = parseNode.Pos;
                    theUnit.CodeLength=nodeParList.Pos-parseNode.Pos+nodeParList.Length;
                    parseNode = new ParseNode(ParseNodeType.FunctionCall, index, i - index, new[] { parseNode, nodeParList });
                    continue;
                }

                (var memberAccess, var nodeMemberAccess, i2) = GetMemberAccess(context, theUnit, i);
                if (i2 > i)
                {
                    i = i2;
                    theUnit = memberAccess;
                    parseNode = new ParseNode(ParseNodeType.MemberAccess, index, i - index, new[] { parseNode, nodeMemberAccess });
                    continue;
                }

                (var kvc, var nodeKvc, i2) = GetSelectKvcExpression(context, i);
                if (i2 > i)
                {
                    i = i2;
                    theUnit = new SelectorExpression
                    {
                        Source = theUnit,
                        Selector = kvc as KvcExpression,
                        CodePos = i,
                        CodeLength = i2 - i
                    };
                    parseNode = new ParseNode(ParseNodeType.Selection, index, i - index, new[] { parseNode, nodeKvc });
                    continue;
                }

                prog = theUnit;
                return new ExpressionBlockResult(prog, parseNode, i);
            } while (true);
        }
    }
}

File: ts/core/src/parser/syntax/FuncScriptParser.GetCallAndMemberAccess.ts
import { ParseContext, ParseNode, ParseNodeType, ParseResult } from "../FuncScriptParser.Main";
import { GetFunctionCallParametersList } from "./FuncScriptParser.GetFunctionCallParametersList";
import { GetKvcExpression } from "./FuncScriptParser.GetKvcExpression";
import { GetMemberAccess } from "./FuncScriptParser.GetMemberAccess";
import { GetUnit } from "./FuncScriptParser.GetUnit";
import { SkipSpace } from "./FuncScriptParser.SkipSpace";

export function GetCallAndMemberAccess(context: ParseContext, index: number): ParseResult {
    let parseNode: ParseNode | null = null;
    const i1 = SkipSpace(context, index).NextIndex;
    let {  ParseNode: tempParseNode,NextIndex: i } = GetUnit(context, i1);
    
    if (i === index) {
        return { ParseNode: null, NextIndex: index };
    }
    
    parseNode = tempParseNode;

    while (true) {
        let {ParseNode: nodeParList, NextIndex: i2 } = GetFunctionCallParametersList(context,  i);
        if (i2 > i) {
            i = i2;
            parseNode = new ParseNode(ParseNodeType.FunctionCall, index, i - index, [parseNode!, nodeParList!]);
            continue;
        }

        let { ParseNode:nodeMemberAccess, NextIndex:i2Member } = GetMemberAccess(context,  i);
        if (i2Member > i) {
            i = i2Member;
            parseNode = new ParseNode(ParseNodeType.MemberAccess, index, i - index, [parseNode!, nodeMemberAccess!]);
            continue;
        }

        let { ParseNode: nodeKvc, NextIndex: i2Kvc } = GetKvcExpression(context,  i);
        if (i2Kvc > i) {
            i = i2Kvc;
            parseNode = new ParseNode(ParseNodeType.Selection, index, i - index, [parseNode!, nodeKvc!]);
            continue;
        }

        return { ParseNode: parseNode, NextIndex: i };
    }
}